plugin-copilot/src/main/java/appland/copilotChat/copilot/CopilotChatSession.java

```diff
@@ -0,0 +1,133 @@
+                    @NotNull List<CopilotChatRequest.Message> messages,
+                    @Nullable Double temperature,
+                    @Nullable Double topP,
+                    @Nullable Integer n) throws IOException {
+        var maxOutputTokens = model.capabilities().limits().get(CopilotModelDefinition.CopilotCapabilityLimit.MaxOutputTokens);
+        var appliedTemp = temperature != null ? temperature : 0.1;
+        var copilotRequest = new CopilotChatRequest(model.id(), messages, maxOutputTokens, true, appliedTemp, topP, n);
+
+        var chatEndpoint = Urls.newFromEncoded(endpoint + "/chat/completions");
+        HttpRequests.post(chatEndpoint.toExternalForm(), "application/json")
+                .isReadResponseOnError(true)
+                .gzip(true)
+                .tuner(connection -> {
+                    applyHeaders(connection, baseHeaders);
+                    connection.setRequestProperty("openai-intent", "conversation-panel");
+                    connection.setRequestProperty("openai-organization", "github-copilot");
+                    connection.setRequestProperty("x-request-id", requestId());
+                })
+                .connect(httpRequest -> {
+                    httpRequest.write(GitHubCopilotService.gson.toJson(copilotRequest));
+                    readServerEvents(httpRequest.getReader(), responseListener);
+                    return null;
+                });
+    }
+
+    private static void readServerEvents(@NotNull BufferedReader response,
+                                         @NotNull CopilotChatResponseListener responseListener) throws IOException {
+        var pendingEventData = new StringBuilder();
+        try (response) {
+            do {
+                var line = response.readLine();
+                if (line == null) {
+                    break;
+                }
+
+                if (line.startsWith("data:")) {
+                    pendingEventData.append(line.substring("data:".length()).trim());
+                } else if (line.isEmpty()) {
+                    var jsonString = pendingEventData.toString();
+                    pendingEventData.setLength(0);
+
+                    var chunkText = jsonString.trim();
+                    if ("[DONE]".equals(chunkText)) {
+                        responseListener.end();
+                        break;
+                    }
+
+                    processSSEChunk(responseListener, chunkText);
+                }
+            } while (true);
+        }
+    }
+
+    private static void processSSEChunk(@NotNull CopilotChatResponseListener responseListener,
+                                        @NotNull String sseData) {
+        var chunk = GitHubCopilotService.gson.fromJson(sseData, CopilotChatCompletionsStreamChunk.class);
+
+        var id = chunk.id();
+        var modelName = chunk.model();
+        var created = chunk.created();
+        for (var choice : chunk.choices()) {
```

Comment:
This isn't right. You only want the first choice. (Although In general with default parameters there will be only one.)

Reply:
Thanks, good point. I'll fix that when I'm back.